File: examplepatches/SysAdminDoc/hushfeed/extensions/tiktok/src/main/java/app/morphe/extension/tiktok/spoof/sim/SpoofSimPatch.java
/*
 * Forked from:
 * https://gitlab.com/ReVanced/revanced-patches/-/blob/main/extensions/tiktok/src/main/java/app/revanced/extension/tiktok/spoof/sim/SpoofSimPatch.java
 */

package app.morphe.extension.tiktok.spoof.sim;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.shared.Utils;
import app.morphe.extension.tiktok.settings.Settings;

@SuppressWarnings("unused")
public class SpoofSimPatch {
    private static boolean isContextNotSet(String fieldSpoofed) {
        if (Utils.getContext() != null) {
            return false;
        }

        Logger.printException(() -> "Context is not yet set, cannot spoof: " + fieldSpoofed, null);
        return true;
    }

    /**
     * Whether TikTok's startup SIM change check should be skipped.
     *
     * <p>That check sends one telemetry event naming the country and carrier before and after
     * the change, how many SIMs the phone reports, and how long since it last looked. What
     * decides that something changed is the subscription id, an int none of the spoofed getters
     * covers, so the event fires on a real SIM swap or a change of which SIM carries data. The
     * country and carrier in it do read as the preset. The SIM count does not: it comes from
     * SubscriptionManager, and the event still reports that the hardware moved. Nothing else in
     * the app reads what the check stores.
     */
    public static boolean shouldSkipSimChangeReport() {
        if (isContextNotSet("simChangeReport")) return false;

        return Settings.SIM_SPOOF.get();
    }

    public static String getCountryIso(String value) {
        if (isContextNotSet("countryIso")) return value;

        if (Settings.SIM_SPOOF.get()) {
            String iso = Settings.SIM_SPOOF_ISO.get();
            // The dialog checks this; a restored backup does not go through the dialog, and
            // TikTok reads the value straight into code that expects a country code.
            if (!app.morphe.extension.tiktok.spoof.region.RegionSpoof.validCountry(iso)) {
                Logger.printInfo(() -> "Ignoring an unusable SIM country: " + iso);
                return value;
            }
            Logger.printDebug(() -> "Spoofing countryIso from: " + value + " to: " + iso);
            return iso;
        }

        return value;
    }

    public static String getOperator(String value) {
        if (isContextNotSet("MCC-MNC")) return value;

        if (Settings.SIM_SPOOF.get()) {
            String mccMnc = mccMnc(Settings.SIMSPOOF_MCCMNC.get());
            if (!validMccMnc(mccMnc)) {
                Logger.printInfo(() -> "Ignoring an unusable MCC/MNC: " + mccMnc);
                return value;
            }
            Logger.printDebug(() -> "Spoofing sim MCC-MNC from: " + value + " to: " + mccMnc);
            return mccMnc;
        }

        return value;
    }

    public static String getOperatorName(String value) {
        if (isContextNotSet("operatorName")) return value;

        if (Settings.SIM_SPOOF.get()) {
            String operator = Settings.SIMSPOOF_OP_NAME.get();
            Logger.printDebug(() -> "Spoofing sim operatorName from: " + value + " to: " + operator);
            return operator;
        }

        return value;
    }

    public static String getMcc(String value) {
        return getMccMncPart(value, true);
    }

    public static String getMnc(String value) {
        return getMccMncPart(value, false);
    }

    /**
     * Five or six digits, which is what a network code is. Callers inside TikTok read the
     * first three as the country and parse them as a number, so anything else is refused
     * rather than handed on: a value typed before this check existed, or restored from a
     * backup, would otherwise reach them as letters.
     */
    public static boolean validMccMnc(String value) {
        String code = mccMnc(value);
        if (code.length() < 5 || code.length() > 6) return false;
        for (int index = 0; index < code.length(); index++) {
            if (code.charAt(index) < '0' || code.charAt(index) > '9') return false;
        }
        return true;
    }

    /**
     * The stored code without the spaces around it. The dialog and the getters below have to
     * read it the same way, or the field accepts a value that is then ignored every time.
     */
    private static String mccMnc(String value) {
        return value == null ? "" : value.trim();
    }

    private static String getMccMncPart(String value, boolean mcc) {
        if (isContextNotSet(mcc ? "cellMcc" : "cellMnc")) return value;
        if (!Settings.SIM_SPOOF.get()) return value;

        String combined = mccMnc(Settings.SIMSPOOF_MCCMNC.get());
        if (!validMccMnc(combined)) return value;

        String replacement = mcc ? combined.substring(0, 3) : combined.substring(3);
        Logger.printDebug(() -> "Spoofing " + (mcc ? "cell MCC" : "cell MNC")
                + " from: " + value + " to: " + replacement);
        return replacement;
    }
}

File: examplepatches/SysAdminDoc/hushfeed/patches/src/main/kotlin/app/morphe/patches/tiktok/misc/spoof/sim/SpoofSimPatch.kt
/*
 * Forked from:
 * https://gitlab.com/ReVanced/revanced-patches/-/blob/main/patches/src/main/kotlin/app/revanced/patches/tiktok/misc/spoof/sim/SpoofSimPatch.kt
 */

package app.morphe.patches.tiktok.misc.spoof.sim

import app.morphe.patcher.extensions.InstructionExtensions.addInstruction
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.addInstructionsWithLabels
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patcher.util.smali.ExternalLabel
import app.morphe.patches.shared.compat.AppCompatibilities
import app.morphe.patches.tiktok.misc.extension.sharedExtensionPatch
import app.morphe.patches.tiktok.misc.settings.SettingsStatusLoadFingerprint
import app.morphe.patches.tiktok.misc.settings.settingsPatch
import app.morphe.util.findMutableMethodOf
import app.morphe.util.getReference
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.Method
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
import com.android.tools.smali.dexlib2.iface.reference.MethodReference
import app.morphe.patches.tiktok.shared.requireLocals

private const val EXTENSION_CLASS_DESCRIPTOR = "Lapp/morphe/extension/tiktok/spoof/sim/SpoofSimPatch;"

@Suppress("unused")
val simSpoofPatch = bytecodePatch(
    name = "SIM spoof",
    description = "Spoofs SIM country and operator information retrieved by TikTok, with country presets for easier setup.",
    default = true,
) {
    dependsOn(
        sharedExtensionPatch,
        settingsPatch,
    )

    compatibleWith(*AppCompatibilities.tiktok4623())

    execute {
        // Checked before the first write: the SIM change task's injection writes v0 and then
        // falls into TikTok's own first instruction, so v0 has to be a local.
        CheckSimChangeTaskFingerprint.method.requireLocals("SIM spoof", 1)

        val replacements = mapOf(
            "getSimCountryIso" to "getCountryIso",
            "getNetworkCountryIso" to "getCountryIso",
            "getSimOperator" to "getOperator",
            "getNetworkOperator" to "getOperator",
            "getSimOperatorName" to "getOperatorName",
            "getNetworkOperatorName" to "getOperatorName",
        )

        val patchesByMethod = linkedMapOf<Method, ArrayDeque<Pair<Int, String>>>()
        classDefForEach { classDef ->
            if (classDef.type.startsWith("Lapp/morphe/extension/")) return@classDefForEach
            for (method in classDef.methods) {
                val implementation = method.implementation ?: continue
                implementation.instructions.forEachIndexed { index, instruction ->
                    if (
                        instruction.opcode != Opcode.INVOKE_VIRTUAL &&
                        instruction.opcode != Opcode.INVOKE_VIRTUAL_RANGE
                    ) return@forEachIndexed

                    val methodReference = instruction.getReference<MethodReference>()
                        ?: return@forEachIndexed
                    val replacement = when {
                        methodReference.definingClass == "Landroid/telephony/TelephonyManager;" ->
                            replacements[methodReference.name]
                        methodReference.definingClass.startsWith("Landroid/telephony/CellIdentity") &&
                            methodReference.name == "getMccString" -> "getMcc"
                        methodReference.definingClass.startsWith("Landroid/telephony/CellIdentity") &&
                            methodReference.name == "getMncString" -> "getMnc"
                        else -> null
                    }
                    if (replacement == null || methodReference.returnType != "Ljava/lang/String;") {
                        return@forEachIndexed
                    }
                    if (implementation.instructions.elementAtOrNull(index + 1)?.opcode == Opcode.MOVE_RESULT_OBJECT) {
                        patchesByMethod.getOrPut(method) { ArrayDeque() }.add(index to replacement)
                    }
                }
            }
        }

        patchesByMethod.forEach { (method, patches) ->
            val mutableMethod = mutableClassDefBy(method.definingClass).findMutableMethodOf(method)
            while (patches.isNotEmpty()) {
                val (index, replacement) = patches.removeLast()
                val resultRegister = mutableMethod.getInstruction<OneRegisterInstruction>(index + 1).registerA

                mutableMethod.addInstructions(
                    index + 2,
                    """
                        invoke-static/range { v$resultRegister .. v$resultRegister }, $EXTENSION_CLASS_DESCRIPTOR->$replacement(Ljava/lang/String;)Ljava/lang/String;
                        move-result-object v$resultRegister
                    """,
                )
            }
        }

        // TikTok's startup SIM check reports a change to its own telemetry: which way the SIM
        // went, how long since the last check, how many SIMs the phone has, and the country and
        // carrier before and after. What decides that something changed is the subscription id,
        // an int the spoof does not touch, so the event fires on a real SIM swap or a change of
        // which SIM carries data. The country and carrier in it read as the preset, but the SIM
        // count does not: it comes from SubscriptionManager rather than TelephonyManager, and
        // the event still says the hardware moved. Nothing else in the app reads what the task
        // stores, so with the preset on it does not run.
        CheckSimChangeTaskFingerprint.method.apply {
            addInstructionsWithLabels(
                0,
                """
                    invoke-static {}, $EXTENSION_CLASS_DESCRIPTOR->shouldSkipSimChangeReport()Z
                    move-result v0
                    if-eqz v0, :run_sim_check
                    return-void
                """,
                ExternalLabel("run_sim_check", getInstruction(0)),
            )
        }

        SettingsStatusLoadFingerprint.method.addInstruction(
            0,
            "invoke-static {}, Lapp/morphe/extension/tiktok/settings/SettingsStatus;->enableSimSpoof()V",
        )
    }
}
